{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMXRYU7pN2UfRlJgmfxbbHB"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["\n","## Blob\n","- 주로 대용량 데이터를 저장할때 사용\n","- `Scale up(Database)`\n"," \n","# File Storage(=port:445, file share)\n","- 윈도우의 D드라이브 같은 것(존재하지 않는 가상의 드라이브)\n","- 파일 공유를 하며 여러 유저가 동시에 파일에 접근하고 데이터를 저장할때 사용\n","-  `Scale out(Web)`\n","\n","----"],"metadata":{"id":"RGFrcHqyiw0P"}},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"OCygBLe28sXW","executionInfo":{"status":"ok","timestamp":1681968925721,"user_tz":-540,"elapsed":9201,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"7abfd879-25bf-4b7e-e7b7-ca66a2c92808"},"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Requirement already satisfied: azure-storage-file in /usr/local/lib/python3.9/dist-packages (2.1.0)\n","Requirement already satisfied: azure-common>=1.1.5 in /usr/local/lib/python3.9/dist-packages (from azure-storage-file) (1.1.28)\n","Requirement already satisfied: azure-storage-common~=2.1 in /usr/local/lib/python3.9/dist-packages (from azure-storage-file) (2.1.0)\n","Requirement already satisfied: cryptography in /usr/local/lib/python3.9/dist-packages (from azure-storage-common~=2.1->azure-storage-file) (40.0.2)\n","Requirement already satisfied: python-dateutil in /usr/local/lib/python3.9/dist-packages (from azure-storage-common~=2.1->azure-storage-file) (2.8.2)\n","Requirement already satisfied: requests in /usr/local/lib/python3.9/dist-packages (from azure-storage-common~=2.1->azure-storage-file) (2.27.1)\n","Requirement already satisfied: cffi>=1.12 in /usr/local/lib/python3.9/dist-packages (from cryptography->azure-storage-common~=2.1->azure-storage-file) (1.15.1)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.9/dist-packages (from python-dateutil->azure-storage-common~=2.1->azure-storage-file) (1.16.0)\n","Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.9/dist-packages (from requests->azure-storage-common~=2.1->azure-storage-file) (1.26.15)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.9/dist-packages (from requests->azure-storage-common~=2.1->azure-storage-file) (2022.12.7)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.9/dist-packages (from requests->azure-storage-common~=2.1->azure-storage-file) (3.4)\n","Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.9/dist-packages (from requests->azure-storage-common~=2.1->azure-storage-file) (2.0.12)\n","Requirement already satisfied: pycparser in /usr/local/lib/python3.9/dist-packages (from cffi>=1.12->cryptography->azure-storage-common~=2.1->azure-storage-file) (2.21)\n"]}],"source":["!pip install azure-storage-file"]},{"cell_type":"code","source":["from azure.storage.file import FileService\n","file_service=FileService(account_name='usercontainer91',account_key='')"],"metadata":{"id":"qtntOnpy9NRy"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["file_service.create_share('myshare') # file share 기본적으로 5GB씩 배정됨"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5RhH7EoOG89_","executionInfo":{"status":"ok","timestamp":1681968933007,"user_tz":-540,"elapsed":1045,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"6a6d1a3e-d7f8-4935-9d49-a7d4beb0d8a3"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["False"]},"metadata":{},"execution_count":3}]},{"cell_type":"code","source":["file_service.create_directory('myshare','sampledir') # mkdir"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"nn79BThBHJNM","executionInfo":{"status":"ok","timestamp":1681968934869,"user_tz":-540,"elapsed":449,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"a9a47a56-7562-4615-bdb4-9de997b9271c"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["False"]},"metadata":{},"execution_count":4}]},{"cell_type":"code","source":["from azure.storage.file import ContentSettings\n","file_service.create_file_from_path(\n","    'myshare',\n","    'sampledir',\n","    'myfile.jpeg',\n","    'mmayo.jpeg'\n",")"],"metadata":{"id":"1wXkdnNbHuHj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# directory 내용의 확인\n","generator =file_service.list_directories_and_files('myshare')\n","for file_or_dir in generator:\n","  print(file_or_dir.name)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"_p5b1ksjJQvn","executionInfo":{"status":"ok","timestamp":1681969269955,"user_tz":-540,"elapsed":411,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"76711703-57f0-4c31-ab00-3bd03aaed33b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["myfile\n","sampledir\n"]}]},{"cell_type":"code","source":["# 파일을 가져오는 코드\n","file_service.get_file_to_path('myshare',\n","                              None,\n","                              'myfile',\n","                              'mmayo_DOWNLOAD.jpeg'\n","                              )"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Mfj3CoQrLEJE","executionInfo":{"status":"ok","timestamp":1681969328555,"user_tz":-540,"elapsed":1452,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"522a2ade-62c7-4103-8f96-49f620d7e9b9"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["<azure.storage.file.models.File at 0x7fc1cc3544c0>"]},"metadata":{},"execution_count":19}]},{"cell_type":"code","source":["# snapshot :사용자가 할 수 있는 백업..?\n","metadata ={'foo':'bar'}\n","snapshot = file_service.snapshot_share('myshare',metadata=metadata)"],"metadata":{"id":"ae9RL9fMLyj4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["share = list(file_service.list_shares(include_snapshots=True))"],"metadata":{"id":"428cQkTbNLNr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#원하는 snapshot으로 복구?\n","directories_and_file = list(file_service.list_directories_and_files('myshare',\n","                                                                    snapshot='2023-04-20T05:48:10.0000000Z'))\n","for file_or_dir in directories_and_file:\n","  print(file_or_dir.name)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kbj3tE4yRs9S","executionInfo":{"status":"ok","timestamp":1681971482577,"user_tz":-540,"elapsed":1433,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"954a2065-b368-4c8a-c016-cdef8547b5b9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["myfile\n","sampledir\n"]}]},{"cell_type":"code","source":["#snapshot 제거\n","file_service.delete_share('myshare',snapshot='2023-04-20T05:48:10.0000000Z')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"VOt9pGCeSK25","executionInfo":{"status":"ok","timestamp":1681971747731,"user_tz":-540,"elapsed":436,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"ee304205-deb5-4eb2-bd44-e430f6607afa"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":29}]},{"cell_type":"code","source":["# delete sharefolder\n","file_service.delete_file('myshare',None,'myfile')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":428},"id":"zmo86kFVVBd4","executionInfo":{"status":"error","timestamp":1681972012493,"user_tz":-540,"elapsed":493,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"b2498ffa-5e4d-4be1-a3b8-75efdb246b79"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:azure.storage.common.storageclient:Client-Request-ID=579ca59e-df44-11ed-9058-0242ac1c000c Retry policy did not allow for a retry: Server-Timestamp=Thu, 20 Apr 2023 06:26:53 GMT, Server-Request-ID=8cefe79a-e01a-00ba-1251-732e48000000, HTTP status code=404, Exception=The specified resource does not exist. ErrorCode: ResourceNotFound<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>ResourceNotFound</Code><Message>The specified resource does not exist.RequestId:8cefe79a-e01a-00ba-1251-732e48000000Time:2023-04-20T06:26:53.9792443Z</Message></Error>.\n"]},{"output_type":"error","ename":"AzureMissingResourceHttpError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAzureMissingResourceHttpError\u001b[0m             Traceback (most recent call last)","\u001b[0;32m<ipython-input-33-ae315e8a6810>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# delete sharefolder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mfile_service\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'myshare'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'myfile'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/file/fileservice.py\u001b[0m in \u001b[0;36mdelete_file\u001b[0;34m(self, share_name, directory_name, file_name, timeout)\u001b[0m\n\u001b[1;32m   1825\u001b[0m         \u001b[0mrequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'timeout'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0m_int_to_str\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1826\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1827\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_perform_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1828\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1829\u001b[0m     def create_file(self, share_name, directory_name, file_name,\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    444\u001b[0m                                  \u001b[0mstatus_code\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    445\u001b[0m                                  exception_str_in_one_line)\n\u001b[0;32m--> 446\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    447\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    448\u001b[0m                 \u001b[0;31m# If this is a location locked operation and the location is not set,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    372\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mAzureException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    373\u001b[0m                     \u001b[0mretry_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 374\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    375\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    376\u001b[0m                     \u001b[0mretry_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    357\u001b[0m                         \u001b[0;31m# This exception will be caught by the general error handler\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    358\u001b[0m                         \u001b[0;31m# and raised as an azure http exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 359\u001b[0;31m                         _http_error_handler(\n\u001b[0m\u001b[1;32m    360\u001b[0m                             HTTPError(response.status, response.message, response.headers, response.body))\n\u001b[1;32m    361\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/_error.py\u001b[0m in \u001b[0;36m_http_error_handler\u001b[0;34m(http_error)\u001b[0m\n\u001b[1;32m    113\u001b[0m     \u001b[0mex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_code\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mAzureMissingResourceHttpError\u001b[0m: The specified resource does not exist. ErrorCode: ResourceNotFound\n<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>ResourceNotFound</Code><Message>The specified resource does not exist.\nRequestId:8cefe79a-e01a-00ba-1251-732e48000000\nTime:2023-04-20T06:26:53.9792443Z</Message></Error>"]}]},{"cell_type":"code","source":["file_service.delete_share('myshare')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":428},"id":"g8DFtpl4Vuno","executionInfo":{"status":"error","timestamp":1681972023774,"user_tz":-540,"elapsed":838,"user":{"displayName":"한승훈","userId":"08150251924661679991"}},"outputId":"2d46c617-8554-4b5b-e221-aee4be1c3f9c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["ERROR:azure.storage.common.storageclient:Client-Request-ID=5e295d9e-df44-11ed-9058-0242ac1c000c Retry policy did not allow for a retry: Server-Timestamp=Thu, 20 Apr 2023 06:27:04 GMT, Server-Request-ID=8cefe844-e01a-00ba-2d51-732e48000000, HTTP status code=409, Exception=The share has snapshots and the operation requires no snapshots. ErrorCode: ShareHasSnapshots<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>ShareHasSnapshots</Code><Message>The share has snapshots and the operation requires no snapshots.RequestId:8cefe844-e01a-00ba-2d51-732e48000000Time:2023-04-20T06:27:04.9696961Z</Message></Error>.\n"]},{"output_type":"error","ename":"AzureConflictHttpError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAzureConflictHttpError\u001b[0m                    Traceback (most recent call last)","\u001b[0;32m<ipython-input-34-65a0b274988b>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfile_service\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete_share\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'myshare'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/file/fileservice.py\u001b[0m in \u001b[0;36mdelete_share\u001b[0;34m(self, share_name, fail_not_exist, timeout, snapshot, delete_snapshots)\u001b[0m\n\u001b[1;32m    976\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    977\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mAzureHttpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 978\u001b[0;31m                 \u001b[0m_dont_fail_not_exist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    979\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    980\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/_error.py\u001b[0m in \u001b[0;36m_dont_fail_not_exist\u001b[0;34m(error)\u001b[0m\n\u001b[1;32m     95\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/file/fileservice.py\u001b[0m in \u001b[0;36mdelete_share\u001b[0;34m(self, share_name, fail_not_exist, timeout, snapshot, delete_snapshots)\u001b[0m\n\u001b[1;32m    973\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mfail_not_exist\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    974\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 975\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_perform_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpected_errors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_SHARE_NOT_FOUND_ERROR_CODE\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    976\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    977\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mAzureHttpError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    444\u001b[0m                                  \u001b[0mstatus_code\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    445\u001b[0m                                  exception_str_in_one_line)\n\u001b[0;32m--> 446\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    447\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    448\u001b[0m                 \u001b[0;31m# If this is a location locked operation and the location is not set,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    372\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mAzureException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    373\u001b[0m                     \u001b[0mretry_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 374\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    375\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    376\u001b[0m                     \u001b[0mretry_context\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/storageclient.py\u001b[0m in \u001b[0;36m_perform_request\u001b[0;34m(self, request, parser, parser_args, operation_context, expected_errors)\u001b[0m\n\u001b[1;32m    357\u001b[0m                         \u001b[0;31m# This exception will be caught by the general error handler\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    358\u001b[0m                         \u001b[0;31m# and raised as an azure http exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 359\u001b[0;31m                         _http_error_handler(\n\u001b[0m\u001b[1;32m    360\u001b[0m                             HTTPError(response.status, response.message, response.headers, response.body))\n\u001b[1;32m    361\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.9/dist-packages/azure/storage/common/_error.py\u001b[0m in \u001b[0;36m_http_error_handler\u001b[0;34m(http_error)\u001b[0m\n\u001b[1;32m    113\u001b[0m     \u001b[0mex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_code\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_code\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mAzureConflictHttpError\u001b[0m: The share has snapshots and the operation requires no snapshots. ErrorCode: ShareHasSnapshots\n<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>ShareHasSnapshots</Code><Message>The share has snapshots and the operation requires no snapshots.\nRequestId:8cefe844-e01a-00ba-2d51-732e48000000\nTime:2023-04-20T06:27:04.9696961Z</Message></Error>"]}]},{"cell_type":"code","source":[],"metadata":{"id":"QK5bV6-WV6vE"},"execution_count":null,"outputs":[]}]}